src/core/mesh/qgsmeshdatasetgroupstore.cpp

```diff
@@ -257,13 +279,37 @@ void QgsMeshDatasetGroupStore::readXml( const QDomElement &storeElem, const QgsR
   Q_UNUSED( context );
   mRegistery.clear();
   QDomElement datasetElem = storeElem.firstChildElement( "mesh-dataset" );
+  QMap<int, QgsMeshDatasetGroup *> extraDatasetGroups;
   while ( !datasetElem.isNull() )
   {
     int globalIndex = datasetElem.attribute( QStringLiteral( "global-index" ) ).toInt();
-    int sourceIndex = datasetElem.attribute( QStringLiteral( "source-index" ) ).toInt();
+    int sourceIndex;
     QgsMeshDatasetSourceInterface *source = nullptr;
     if ( datasetElem.attribute( QStringLiteral( "source-type" ) ) == QStringLiteral( "persitent-provider" ) )
+    {
       source = mPersistentProvider;
+      sourceIndex = datasetElem.attribute( QStringLiteral( "source-index" ) ).toInt();
+    }
+    else
+    {
+      QString keyGenerator = datasetElem.attribute( QStringLiteral( "source-type" ) );
+      source = mExtraDatasets.get();
+      QString name = datasetElem.attribute( QStringLiteral( "name" ) );
+      QString formula = datasetElem.attribute( QStringLiteral( "formula" ) );
+      qint64 startTime = datasetElem.attribute( QStringLiteral( "startTime" ) ).toLongLong();
+      qint64 endTime = datasetElem.attribute( QStringLiteral( "endTime" ) ).toLongLong();
+      if ( QgsApplication::instance() )
+      {
+        QgsMeshDataGeneratorInterface *generator =
+          QgsApplication::instance()->meshDataGeneratorRegistry()->meshDataGenerator( keyGenerator );
+        if ( generator )
+        {
+          QgsMeshDatasetGroup *dsg = generator->createVirtualDatasetGroupFromFormula( name, formula, mLayer, startTime, endTime );
+          extraDatasetGroups[globalIndex] = dsg;
+          sourceIndex = mExtraDatasets->addDatasetGroup( dsg );
+        }
+      }
```

Comment:
I do not understand why we need this `QgsMeshDataGeneratorInterface` and associated registry. At the moment, it is a lot of infrastructure, and at the end, all it does it just create QgsVirtualDatasetGroup with default contructor. I suggest to remove all this QgsMeshDataGenerator* infrastructure and add it only if/when needed. I do not see clearly who/how/why want to create custom generators...

Reply:
as the virtual dataset group is in the analysis API (because dependent of the mesh calculator), and we need to create virtual dataset group in the core (when reading xml in mesh layer).


> who/how/why want to create custom generators...

us :-)
Indeed, this could open new possibility, for example editing/creation of dataset/mesh in the analysis and bring this possibility indirectly in the core with interface.

Reply:
ok